### PR TITLE
replace RequestHeaderMap in tracers with general TraceContext

### PR DESCRIPTION
**Pull request by wbpcode**
Sub-PR of #16049. Check #16049 get more background information.

This PR is last sub-PR of #16049. It simply replaces `Http::RequestHeaderMap` with general `Tracing::TraceContext` in all tracer drivers implementations.

Check #16793 get more info about `Tracing::TraceContext`.

After this PR, the main body of the entire general tracing system is completed. Next, we can try to use the new general tracing system in dubbo and thrift.

Commit Message: replace RequestHeaderMap in tracers with general TraceContext
Risk Level: Low.
Testing: N/A.
Docs Changes: N/A.
Release Notes: N/A.

### CODE REVIEW

File: source/extensions/tracers/common/ot/opentracing_driver_impl.cc [hunk 1]
```diff
@@ -21,52 +21,57 @@ Http::RegisterCustomInlineHeader<Http::CustomInlineHeaderRegistry::Type::Request
     ot_span_context_handle(Http::CustomHeaders::get().OtSpanContext);
 
 namespace {
-class OpenTracingHTTPHeadersWriter : public opentracing::HTTPHeadersWriter {
+class OpenTracingHeadersWriter : public opentracing::HTTPHeadersWriter {
```
[the remark this answers is not in the document]
Comment (wbpcode) [reply source/extensions/tracers/common/ot/opentracing_driver_impl.cc:24]: 🤔 I think there is no problem, because here only simple set and get operations are needed on TraceContext. So it is completely possible to use TraceContext to replace HeaderMap.
TraceContext does not support iterative API for the time being, so dynamic_cast is used to temporarily bypass this problem.
And this is an overall change, we can't just keep this one class without making changes.

File: source/common/tracing/http_tracer_impl.cc [hunk 2]
```diff
@@ -317,12 +317,12 @@ RequestHeaderCustomTag::RequestHeaderCustomTag(
...
 absl::string_view RequestHeaderCustomTag::value(const CustomTagContext& ctx) const {
-  if (!ctx.request_headers) {
+  if (ctx.trace_context == nullptr) {
     return default_value_;
   }
   // TODO(https://github.com/envoyproxy/envoy/issues/13454): Potentially populate all header values.
-  const auto entry = ctx.request_headers->get(name_);
-  return !entry.empty() ? entry[0]->value().getStringView() : default_value_;
+  const auto entry = ctx.trace_context->getTraceContext(name_);
+  return entry.has_value() ? entry.value() : default_value_;
```
Comment (lizan) [inline source/common/tracing/http_tracer_impl.cc:325]:
```suggestion
  return entry.value_or(default_value_);
```

File: source/extensions/tracers/common/ot/opentracing_driver_impl.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (lizan) [reply source/extensions/tracers/common/ot/opentracing_driver_impl.cc:24]: The superclass is HTTPHeadersWriter so this is still HTTP only?

File: source/common/tracing/http_tracer_impl.cc (hunk 2 above)
Comment (wbpcode) [reply source/common/tracing/http_tracer_impl.cc:325]: Get it.

File: source/extensions/tracers/common/ot/opentracing_driver_impl.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (wbpcode) [reply source/extensions/tracers/common/ot/opentracing_driver_impl.cc:24]: @lizan Apart from the name, `HTTPHeadersWriter` does not seem to have anything related to the HTTP protocol. 

I have checked source code of opentracing interface and also implementation of `datadog/lightstep`. `HTTPHeadersWriter` does not contain any HTTP-dependent interfaces. `HTTPHeadersWriter` simply inherits from `TextMapWriter`, without adding any HTTP-related methods. 

```
class TextMapReader {
 public:
  virtual ~TextMapReader() = default;
  virtual expected<string_view> LookupKey(string_view /*key*/) const {
    return make_unexpected(lookup_key_not_supported_error);
  }
  virtual expected<void> ForeachKey(
      std::function<expected<void>(string_view key, string_view value)> f)
      const = 0;
};

class TextMapWriter {
 public:
  virtual ~TextMapWriter() = default;
  virtual expected<void> Set(string_view key, string_view value) const = 0;
};

class HTTPHeadersReader : public TextMapReader {};
class HTTPHeadersWriter : public TextMapWriter {};
```
`datadog/lightstep` also does not provide any operations for `HTTPHeadersWriter` that are different from `TextMapWriter`.

In summary, `HTTPHeadersWriter` will not affect the versatility of tracers.
We can completely replace it with `TextMapWriter`. The reason why I have not done this at present is because I don't want the general tracing system to have any impact on existing HTTP Tracers users.

**Review by lizan — APPROVED** [review]
